text/unification.md

```diff
@@ -0,0 +1,268 @@
+- `B` is its type (`evar_concl`)
+
+To be used in a term, existential variables need to be applied to a substitution of its context. So, they take the form `?e@{a₁:=t₁,...,aₙ:tₙ}` where `a₁:=t₁,...,aₙ:=tₙ` is a *substitution* (sometimes called *instance*, often written `args` in the code).
+
+## Existential variables map
+
+A collection of declarations of existential variables is conventionally written `Σ` (an `evar_map` in the code, using either `sigma` or `evd` to refers to it).
+
+## Unification problems
+
+Given an existential variable map, *unification problems* are equations of the form `Γ ⊢ t₁ =?= t₂` or inequations of the form `Γ ⊢ t₁ ≤? t₂` where `t₁` and `t₂` are terms possibly mentioning existential variables with their substitution and both terms living in a context of variables `Γ` (that is the juxtaposition of a `named_context` and of a `rel_context`). An inequation corresponds to a subtyping problem between types (e.g. `Prop ≤ Type`). Note that we assume the existence of a global context of inductive types, definitions and axioms common to the definitions of existential variables and to the unification problems.
+
+A unification algorithm turns a `Σ` and a collection of unification problems into either an impossibility of satisfaction of the equations, or in a collection of (well-typed) definitions for the existential variables such that all problems are satisfied, or a residual set of unification problems that the algorithm is neither able to invalidate nor to satisfy (since unification in general is undecidable in the Calculus of Inductive Constructions).
+
+## Reflexive unification problems
+
+We call *reflexive unification problem* an equation of the form `Γ ⊢ ?e@{subst} =?= ?e@{subst'}`.
+
+## Atomic unification problems
+
+We call *atomic unification problem* an equation of the form `Γ ⊢ ?e@{subst} =?= ?e'@{subst'}` or an inequation of the form `Γ ⊢ ?e@{subst} ≤? ?e'@{subst'}`.
+
+## Simple unification problems
+
+We call *simple unification problem* an equation of the form `Γ ⊢ ?e@{subst} =?= t` or an inequation of the form `Γ ⊢ ?e@{subst} ≤? t`. Unification basically reduces arbitrary complex unification problems to simple ones which in turn may generate new arbitrary complex problems.
+
+## Simplification of reflexive unification problems
+
+A reflexive unification problem induces constraints on the possible definitions of `?e`: once `?e` is defined, it may generate new unification problems between terms of the instances that contain existential variables.
+
+A reflexive unification problem can be simplified as follows:
+- Variables of the context whose instances in the substitution are provably incompatible (e.g. because corresponding to distinct head normal forms) can be removed from the context.
+- If both instances are free of existential variables, the problem can be dropped since its resolution will eventually only create equations between convertible existential-variables-free terms.
```

Comment:
I'm not sure I understand this one.  Is conversion run again after unification completes?  How do you get convertibility constraints?  Also, this seems overly optimistic, consider `x : nat |- ?e : nat` and `|- ?f : nat` and the pair of unification problems `|- ?e@{x:=O} == ?e@{x:=S O}` and `x : nat |- ?e@{x:=x} == match x with O => O | S _ => ?f end`.  We must instantiate `?e` with `match x with O => O | S _ => ?f end`, and only after this substitution do we get `|- O == ?f` out of the first unification problem, which it seems would have been dropped according to this rule.

It seems that Coq can solve this problem, though?
```coq
Check let e := (fun x => ?[e]) in
      let f := ?[f] in
      eq_refl : (e, e O) = (fun x => match x with O => O | S _ => f end, e (S O)).
```

Reply:
It seems I meant "If both instances are pointwise existential-variable-free compatible terms". So, with your previous remark, it should be "If both instances are pointwise existential-variable-free compatible neutral terms".

Your example works because `e := fun x => match x with O => O` is solved first. If you invert the order of unification problems:
```
Check let e := (fun x => ?[e]) in
      let f := ?[f] in
      eq_refl : (e O, e) = (e (S O), fun x => match x with O => O | S _ => f end).
```
it fails.